### PR TITLE
adds link to service provider of @u forwarded email

### PR DESCRIPTION
**Pull request by mikeseibel**

### CODE REVIEW

File: myuw/data/forward_mapping.csv [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+msn.com,https://www.msn.com,MSN,fa-windows
+aol.com,https://my.screenname.aol.com,AOL,fa-envelope
+live.com,https://mail.live.com/m,MS Live,fa-envelope
+163.com,https://mail.163.com/,NetEase,fa-envelope
+outlook.com,https://www.outlook.com,Outlook,fa-windows
+earthlink.net,https://webmail.earthlink.net,EarthLink,fa-envelope
+mac.com,https://www.icloud.com,Mac Mail,fa-apple
+me.com,https://www.icloud.com,Mobile Me,fa-apple
+126.com,https://mail.126.com,NetEase,fa-envelope
+qq.com,https://en.mail.qq.com/cgi-bin/loginpage,QQMail,fa-envelope
```
Comment (fanglinfang) [inline myuw/data/forward_mapping.csv]: QQ email url can be shorter:
https://en.mail.qq.com/

File: myuw/data/forward_mapping.csv [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+deskmail.washington.edu,https://alpine.washington.edu,UW Deskmail,fa-envelope
+gamail.uw.edu,https://mail.google.com,UW Gmail,fa-google
+exchange.washington.edu,https://outlook.office365.com,UW Office 365,fa-windows
+gmail.com,https://mail.google.com,GMail,fa-google
+hotmail.com,https://mail.live.com/m,Hotmail,fa-windows
+yahoo.com,https://mail.yahoo.com,Yahoo,fa-yahoo
+comcast.net,https://login.comcast.net,Comcast,fa-envelope
+msn.com,https://www.msn.com,MSN,fa-windows
+aol.com,https://my.screenname.aol.com,AOL,fa-envelope
+live.com,https://mail.live.com/m,MS Live,fa-envelope
```
Comment (fanglinfang) [inline myuw/data/forward_mapping.csv]: These two works equally well for me:
https://outlook.live.com/owa/
https://mail.live.com/

File: myuw/dao/emaillink.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+This class encapsulates the mapping between
+email address and service provider login
+"""
+
+import re
+import os
+import csv
+
+
+class EmailServiceUrlException(Exception):
```
Comment (vegitron) [inline myuw/dao/emaillink.py]: I like having a custom exception, but there's a general convention (used at least in restclients) to put them in exceptions.py.  I'd like to use that convention in myuw.

File: myuw/test/dao/emaillink.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+             ("https://www.icloud.com", "iCloud Mail", "fa-apple")),
+        ]
+
+        for address in addresses:
+            url, title, icon = get_service_url_for_address(
+                address[0])
+        self.assertEquals(url, address[1][0])
+        self.assertEquals(title, address[1][1])
+        self.assertEquals(icon, address[1][2])
+
```
Comment (vegitron) [inline myuw/test/dao/emaillink.py]: What happens if someone passes in a non-string type?

File: myuw/dao/emaillink.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    path = os.path.join(
+        os.path.dirname(__file__),
+        '..', 'data', 'forward_mapping.csv')
+
+    with open(path, 'rbU') as csvfile:
+        reader = csv.reader(csvfile, delimiter=',', quotechar='"')
+        # skip header
+        next(reader)
+        for row in reader:
+            if re.match('.*[@\.]%s$' % row[0], address, re.I):
```
Comment (vegitron) [inline myuw/dao/emaillink.py]: The case insensitivity isn't tested - is it required?

File: myuw/views/api/uwemail.py [hunk 6]
```diff
@@ -25,6 +27,18 @@ def GET(self, request):
         try:
             my_uwemail_forwarding = get_email_forwarding_for_current_user()
             resp_json = my_uwemail_forwarding.json_data()
+            try:
+                (resp_json['login_url'],
+                 resp_json['title'],
+                 resp_json['icon']) = get_service_url_for_address(
+                     my_uwemail_forwarding.fwd)
+            except EmailServiceUrlException:
```
Comment (vegitron) [inline myuw/views/api/uwemail.py]: This fallback isn't being tested

File: myuw/dao/emaillink.py (hunk 3 above)
Comment (mikeseibel) [inline myuw/dao/emaillink.py]: done

File: myuw/dao/emaillink.py (hunk 5 above)
Comment (mikeseibel) [inline myuw/dao/emaillink.py]: domains aren't case sensitive.  tests adjusted

File: myuw/test/dao/emaillink.py (hunk 4 above)
Comment (mikeseibel) [inline myuw/test/dao/emaillink.py]: guess some test has to break that ground.  done.

File: myuw/views/api/uwemail.py (hunk 6 above)
Comment (mikeseibel) [inline myuw/views/api/uwemail.py]: ...overtaken by events

**Comment by vegitron** [comment]
Once the documented mock data is in place I'll sign off.

File: myuw/data/forward_mapping.csv [hunk 7]
```diff
@@ -0,0 +1,19 @@
+domain,login_url,link_title,icon_style
+deskmail.washington.edu,https://alpine.washington.edu,UW Deskmail,fa-envelope
+gamail.uw.edu,https://mail.google.com,UW Gmail,fa-google
```
Comment (vegitron) [inline myuw/data/forward_mapping.csv]: https://gmail.google.com will pre-select the uw login account - this could open any of their google acccounts

Comment (fanglinfang) [inline myuw/data/forward_mapping.csv]: Mike, you may want to use:
https://mail.google.com/a/uw.edu

**Pull request merged**